Fix fire print on Firefox

KarolDawidziuk opened the pull request:
## What is the purpose of this pull request?

Bug fix 

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [ ] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#<4790>](https://github.com/ckeditor/ckeditor4/issues/4790): Fixed: Incorrect fire Print on FF
```

## What changes did you make?

The problem was with the page load status in FF mode, because an attempt to print was triggered when the preview page was not ready. So I added a `onreadystatechange` callback to wait for the `complete` status.

## Which issues does your PR resolve?

Closes [#4790](https://github.com/ckeditor/ckeditor4/issues/4790).

Comandeer commented on plugins/preview/plugin.js:
```diff
@@ -155,6 +158,14 @@
+				return '';
+			}
+
+			return '<script> window.onreadystatechange = firePrint </script>';
```
It should be bound to `document`, because `readystatechange` event does not bubble up to `window`.

```suggestion
			return '<script> document.onreadystatechange = firePrint; </script>';
```

Comandeer reviewed (changes requested):
I'm thinking about refactoring the whole thing a little bit and moving most of its part into `preview` plugin (see my inline comments).

Additionally there are two failing tests inside `print` suite (the one you modified and one about immediately invoking print on `document.readyState === 'complete'`). However, instead of fixing, we should probably remove them and instead add some tests to `preview` plugin to check if passed calback has been executed in the right time.

KarolDawidziuk commented:
Thanks, @Comandeer for refactoring and review. 
It's ready for another one.

Changes:
* update docs,
* revert old version of unit test from `tests/../print.js`,
* removed two unit tests from `tests/../print.js`,
* added unit test.

Comandeer reviewed (changes requested):
Unfortunately our current approach does not work correctly in IE as it does not change `document.readyState` to `complete` if there are any images in the content. Additionally simply switching to `load` event instead also won't work, because Safari does not fire `load` for subsequent opening of preview window. It means that we have to introduce two different flows – one for IE, using (probably) `load` event and other one for all other browsers.

Comandeer commented:
Blocked by #4824.

KarolDawidziuk commented:
@Comandeer, I've added two ways to invoke the `print` callback. One for IE with using `onload` and `onreadystatechange` for all other browsers.

Comandeer commented on plugins/preview/plugin.js:
```diff
@@ -74,27 +78,45 @@
+					}
+				};
+
+				nativePreviewWindow.fireCallback();
```
Invoking it here means that in IE callback is invoked immediately, without waiting for `load` event (due to the second `if`).

Probably we will need to duplicate check of `readyState` outside of `fireCallback`.

Comandeer reviewed (changes requested):
Overally looks good, there's only one issue with the moment of invoking callback in IE (see inline comment).

Please also rebase the PR onto latest `master`.

KarolDawidziuk commented:
Thanks, @Comandeer for the review.

Now it's working properly for me on every browser, and it's ready for another review.  
Also, due to the rollback of my previous changes related to the CDN opening "preview"  problem  [#4444](https://github.com/ckeditor/ckeditor4/issues/4444) I am adding these changes to this PR.

Comandeer reviewed (changes requested):
Fix seems to work well, good job!

However there are still some small issues:

* see my inline comments;
* there is a mismatch between manual test `.md` and `.html` file, resulting in non-functional test – please fix it;
* it'd be good to add two more unit tests – one for IE and checking if callback is really fired on `load` and one for other browsers to check if it's fired on `readystatechange` with appropriate `readyState`;
* and traditionally – rebase onto latest `master`.

KarolDawidziuk commented on plugins/preview/plugin.js:
```diff
@@ -155,6 +174,21 @@
+			// On IE onreadystatechange does not change document.readyState to complete if there are any images in the content.
+			// So we need introduce a two flows. One for IE and second for all other browsers. (#4790)
+			if ( CKEDITOR.env.ie ) {
+				return '<script> window.focus(); window.onload = function() { fireCallback() }; </script>';
```
> Is it focusing window necessary? Fix seems to work correctly after deleting it.

I have cases where it doesn't work without it, but now after ~20 tests on 3 VMs looks like it's working without it, so I'll delete it.
